crates/hir-def/src/body/lower.rs

```diff
@@ -92,6 +103,12 @@ struct ExprCollector<'a> {
     is_lowering_assignee_expr: bool,
     is_lowering_coroutine: bool,
 
+    /// Legacy (`macro_rules!`) macros can have multiple definitions and shadow each other,
+    /// and we need to find the current definition. So we track the number of definitions we saw.
+    current_block_legacy_macro_defs_count: FxHashMap<Name, usize>,
```

Comment:
I wonder, does this work properly across blocks (that is shadowing in different nesting levels? Feels like it should but just double checking here. Aside, we still need to figure out something similar for def map name resolution, though macro resolution in there is generally too messy right now

Reply:
Yes, because each block will have a new defmap. I added a test.